### PR TITLE
[MVP-1649] implement polling for intercom chat

### PR DESCRIPTION
**Pull request by marukohao**
implement polling for intercom chat
update intercom chat flow, when users come back to use intercom, bring them directly to the chat view instead of to the start chat view.

Test:
Untitled1

Story:
https://notifi.atlassian.net/browse/MVP-1649

### CODE REVIEW

File: packages/notifi-react-card/lib/components/NotifiEmailInput.tsx [hunk 1]
```diff
@@ -58,6 +58,7 @@ export const NotifiEmailInput: React.FC<NotifiEmailInputProps> = ({
   return (
     <>
       {intercomView ? (
+        email &&
```
Comment (juni-notifi) [inline packages/notifi-react-card/lib/components/NotifiEmailInput.tsx:61]: why is this necessary?

File: packages/notifi-react-card/lib/components/NotifiTelegramInput.tsx [hunk 2]
```diff
@@ -40,6 +40,7 @@ export const NotifiTelegramInput: React.FC<NotifiTelegramInputProps> = ({
   return (
     <>
       {intercomView ? (
+        telegramId &&
```
Comment (juni-notifi) [inline packages/notifi-react-card/lib/components/NotifiTelegramInput.tsx:43]: why is this necessary?

File: packages/notifi-react-card/lib/components/intercom/IntercomCard.tsx [hunk 3]
```diff
@@ -155,6 +173,9 @@ export const IntercomCard: React.FC<
         </>
       );
       break;
+    case 'loadingView':
+      view = null;
```
Comment (juni-notifi) [inline packages/notifi-react-card/lib/components/intercom/IntercomCard.tsx:177]: would be good to have a loading state somehow! maybe as a follow-up

File: packages/notifi-react-card/lib/hooks/useIntercomChat.ts [hunk 4]
```diff
@@ -52,6 +52,32 @@ export const useIntercomChat = ({
...
+            if (nodes[0].id === chatMessages[0].id) {
+              return;
+            } else {
+              nodes.forEach((node, index) => {
+                if (node.id === chatMessages[0].id) {
+                  const newMessages = nodes.slice(0, index);
+                  setChatMessages([...newMessages, ...chatMessages]);
+                }
+              });
+            }
```
Comment (juni-notifi) [inline packages/notifi-react-card/lib/hooks/useIntercomChat.ts:74]: it's probably better to always add all the messages and dedupe on `id`, sorted by message created time.

This logic seems reliant on all new messages being "newer" than the last known newest message, which feels like it could run into edge cases with client-side insert of sendMessage results.

**Review by juni-notifi — COMMENTED** [review]
lmk if you'd like to talk about the algorithm for merging messages!

File: packages/notifi-react-card/lib/hooks/useIntercomChat.ts (hunk 4 above)
Comment (marukohao) [reply packages/notifi-react-card/lib/hooks/useIntercomChat.ts:74]: I am adding the chatMessages as dependency here to avoid the edge case for sending messages, but yeah, it's still insecure. Let me redo the logic as you suggested 👍

File: packages/notifi-react-card/lib/components/NotifiEmailInput.tsx (hunk 1 above)
Comment (marukohao) [reply packages/notifi-react-card/lib/components/NotifiEmailInput.tsx:61]: I was thinking of if there is no email address showing here, we don't need to show the verification link, hmm, we could probably just use the emailIdThatNeedsConfirmation to check it. Let me remove the email here.

File: packages/notifi-react-card/lib/hooks/useIntercomChat.ts [hunk 5]
```diff
@@ -52,6 +55,35 @@ export const useIntercomChat = ({
...
+        .getConversationMessages({
+          first: 50,
+          getConversationMessagesInput: { conversationId },
+        })
+        .then((response) => {
+          if (Array.isArray(response.nodes)) {
+            const nodes = response.nodes;
+            if (nodes[0].id === chatMessages[0].id) {
+              return;
+            } else {
```
Comment (juni-notifi) [inline packages/notifi-react-card/lib/hooks/useIntercomChat.ts:70]: actually i don't think we need to have this conditional

File: packages/notifi-react-card/lib/hooks/useIntercomChat.ts [hunk 6]
```diff
@@ -52,6 +55,35 @@ export const useIntercomChat = ({
...
+              const chatMessageIds = chatMessages.map((message) => message.id);
+              const dedupeNewMessages = nodes.filter(
+                (node) => !chatMessageIds.includes(node.id),
+              );
```
Comment (juni-notifi) [inline packages/notifi-react-card/lib/hooks/useIntercomChat.ts:74]: it would be nice if chatMessageIds were a Set instead (using `reduce` instead of `map`) to improve runtime of `includes` check

**Review by juni-notifi — APPROVED** [review]
thanks for the update!
